### DISCUSSION TITLE
How to access PID power value inside script?

### DISCUSSION THREAD

**Discussion by degreal** [arendst/Tasmota#17749] [Q&A] [answered]
I compiled firmware with:
```
#define USE_SCRIPT
#define USE_PID
#define PID_DONT_USE_PID_TOPIC
```
I see PID algorihm working:

> 00:51:20.408 RSL: SENSOR = {"Time":"2023-01-20T00:51:20","PID":{"PidPv":0.00,"PidSp":19.50,"PidPb":5.00,"PidTi":1800.00,"PidTd":15.00,"PidInitialInt":0.50,"PidDSmooth":3.00,"PidAuto":1,"PidManualPower":0.00,"PidMaxInterval":300,"PidUpdateSecs":5,"PidPower":0.00}}
> 00:51:27.478 RSL: PID = {"power":"0.300"}
> 00:51:28.390 RSL: PID = {"power":"0.299"}

In which script section and how I access the "power" value so I can make further calculations and procedures in my script?

**Comment by gemu2015** [comment, marked as answer]
JSON strings may be evaluated in the >T or >E section depending on the source.

example:
```
>D
power=0
>T
power=PID#PidPower
>S
print %power%
```

**Reply by degreal** [reply]
This only updates value in interval of teleperiod (eg. every 60 seconds) but PID is running faster (5 seconds).
This works somewhat better but how I don't know how to synchronize my procedures with PID arrival event:
```
>D
p=0
>E
p=power
>S
print %p%
```
I would run procedures in "E" section so it would be responsive as possible but how do you filter out if PID message arrived or something else I can filter out?

**Reply by gemu2015** [reply]
PID power is updated every second
i have no device with PID so can't test myself

try  with 

```
>E
power=PID#power
if upd[power]>0 {
; power was updated
}
```

**Reply by degreal** [reply]
Thank you! Last example works as intended!

For clarification USE_SCRIPT_STATUS is also needed and "PID#Power" does not work in "E" section. Working example is:
```
>D
p=0
>E
p=power
if upd[power]>0 {
  power was updated: %p%
}
```
